README.md
VMware Terraform provider for vRealize Automation 7
===
[![Build Status](https://travis-ci.org/vmware/terraform-provider-vra7.svg?branch=master)](https://travis-ci.org/vmware/terraform-provider-vra7)

A self-contained deployable integration between Terraform and vRealize Automation (vRA) which allows Terraform users to request/provision entitled vRA catalog items using Terraform. Supports Terraform destroying vRA provisioned resources.

# Getting Started
These instructions will get you a copy of the project up and run on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

# Prerequisites
To get the vRA plugin up and running you need the following things.
* [Terraform 0.9 or above](https://www.terraform.io/downloads.html)
* [Go Language 1.11.4 or above](https://golang.org/dl/)

# Project Setup
Setup a GOLang project structure as follows:
```
|-/home/<USER>/TerraformPluginProject/
    |-bin/
    |-pkg/
    |-src/

```

# Environment Setup
Set following environment variables.

## Linux Users
`GOROOT` is a golang library path.
```
export GOROOT=/usr/local/go
```

`GOPATH` is a path pointing toward the source code directory.
```
export GOPATH=/home/<user>/TerraformPluginProject
```

`GO111MODULE` is a golang module mode flag (outside of `GOPATH`, you do not need to set `GO111MODULE` to activate module mode).
```
export GO111MODULE=on
```

## Windows Users
`GOROOT` is a golang library path.
```
set GOROOT=C:\Go
```

`GOPATH` is a path pointing toward the source code directory.
```
set GOPATH=C:\TerraformPluginProject
```

`GO111MODULE` is a golang module mode flag (outside of `GOPATH`, you do not need to set `GO111MODULE` to activate module mode).
```
set GO111MODULE=on
```

# Set terraform provider

## Linux Users
Create `~/.terraformrc` and put following content in it.
```
    providers {
         vra7 = "/home/<USER>/TerraformPluginProject/bin/terraform-provider-vra7"
    }
```

## Windows Users
Create `%APPDATA%/terraform.rc` and put following content in it.
```
    providers {
         vra7 = "C:\\TerraformPluginProject\\bin\\terraform-provider-vra7.exe"
    }
```


# Installation
Clone repo code into go project using *go get*
```
    go get github.com/vmware/terraform-provider-vra7

```

# Create Binary

## Linux and MacOS Users
Navigate to `/home/<USER>/TerraformPluginProject/src/github.com/vmware/terraform-provider-vra7` and run go build command to generate plugin binary.
```
    go build -o /home/<USER>/TerraformPluginProject/bin/terraform-provider-vra7

```

## Windows Users
Navigate to `C:\TerraformPluginProject\src\github.com\vmware\terraform-provider-vra7` and run go build command to generate plugin binary.
```
    go build -o C:\TerraformPluginProject\bin\terraform-provider-vra7.exe

```

# Usage

## Configure
The VMware vRA terraform configuration file contains two objects.

### Provider
This part contains service provider details.

Provider block contains four mandatory fields:
* `username` - vRA portal username
* `password` - vRA portal password
* `tenant` - vRA portal tenant
* `host` - End point of REST API
* `insecure` - In case of self-signed certificates, default value is false

**Example:**
```
    provider "vra7" {
      username = "vRAUser1@vsphere.local"
      password = "password123!"
      tenant = "corp.local.tenant"
      host = "http://myvra.example.com/"
      insecure = false
    }

```

### Resource
This part contains any resource that can be deployed on that service provider.
For example, in our case machine blueprint, software blueprint, complex blueprint, network, etc.

**Syntax:**
```
resource "vra7_deployment" "<resource_name1>" {
}
```

The resource block contains mandatory and optional fields.

**Mandatory:**

One of catalog_item_name or catalog_item_id must be specified in the resource configuration.
* `catalog_item_name` - catalog_item_name is a field which contains valid catalog item name from your vRA
* `catalog_item_id` - catalog_item_id is a field which contains a valid catalog item id from your vRA

**Optional:**
* `description` - This is an optional field. You can specify a description for your deployment.
* `reasons` - This is an optional field. You can specify the reasons for this deployment.
* `businessgroup_id` - This is an optional field. You can specify a different Business Group ID from what provided by default in the template request, provided that your account is allowed to do it.
* `businessgroup_name` - This is an optional field. You can specify a different Business Group name from what provided by default in the template request, provided that your account is allowed to do it.
* `count` - This field is used to create replicas of resources. If count is not provided then it will be considered as 1 by default.
* `deployment_configuration` - This is an optional field. It can be used to specify deployment level properties like _leaseDays, _number_of_instances or any custom properties of the deployment. Key is any field name of catalog item and value is any valid user input to the respective field..
* `resource_configuration` - This is an optional field. If blueprint properties have default values or no mandatory property value is required then you can skip this field from terraform configuration file. This field contains user inputs to catalog services. Value of this field is in key value pair. Key is service.field_name and value is any valid user input to the respective field.
* `wait_timeout` - This is an optional field with a default value of 15. It defines the time to wait (in minutes) for a resource operation to complete successfully.


**Example 1:**
```
resource "vra7_deployment" "example_machine1" {
  catalog_item_name = "CentOS 6.3"
  reasons      = "I have some"
  description  = "deployment via terraform"
   resource_configuration = {
         Linux.cpu = "1"
         Windows2008R2SP1.cpu =  "2"
         Windows2012.cpu =  "4"
         Windows2016.cpu =  "2"
     }
     deployment_configuration = {
         _leaseDays = "5"
     }
     count = 3
}
```

**Example 2:**
```
resource "vra7_deployment" "example_machine2" {
  catalog_item_id = "e5dd4fba7f96239286be45ed"
   resource_configuration = {
         Linux.cpu = "1"
         Windows2008.cpu =  "2"
         Windows2012.cpu =  "4"
         Windows2016.cpu =  "2"
     }
     count = 4
}

```

Save this configuration in `main.tf` in a path where the binary is placed.

## Execution
These are the Terraform commands that can be used for the vRA plugin:
* `terraform init` - The init command is used to initialize a working directory containing Terraform configuration files.
* `terraform plan` - Plan command shows plan for resources like how many resources will be provisioned and how many will be destroyed.
* `terraform apply` - apply is responsible to execute actual calls to provision resources.
* `terraform refresh` - By using the refresh command you can check the status of the request.
* `terraform show` - show will set a console output for resource configuration and request status.
* `terraform destroy` - destroy command will destroy all the  resources present in terraform configuration file.

Navigate to the location where `main.tf` and binary are placed and use the above commands as needed.

# Scripts

**update_resource_state.sh**

There are few changes in the way the terraform config file is written.
1. The resource name is renamed to vra7_deployment from vra7_resource.
2. catalog_name is renamed to catalog_item_name and catalog_id is renamed to catalog_item_id.
3. General properties of deployment like description and reasons are to be specified at the resource level map instead of deployment_configuration.
4. catalog_configuration map is removed.
5. Custom/optional properties of deployment are to be specified in deployment_configuration instead of catalog_configuration.

These changes in the config file will lead to inconsistency in the `terraform.tfstate` file of the existing resources provisioned using terraform. 
The existing state files can be converted to the new format using the script, `update_resource_state.sh` under the scripts folder. 

Note: This script will only convert the state file. The changes to the config file(.tf file) still needs to be done manually.

## How to use the script

1. Copy the script, `script/update_resource_state.sh` in the same directory as your terraform.tfstate file.
2. Change the permission of the script, for example `chmod 0700 update_resource_state.sh`.
3. Run the script, `./update_resource_state.sh`.
4. The terraform.tfstate will be updated to the new format and a back-up of the old file is saved as terraform.tfstate_back

# Contributing
The `terraform-provider-vra7` project team welcomes contributions from the community. Before you start working with `terraform-provider-vra7`, please read our [Developer Certificate of Origin](https://cla.vmware.com/dco). All contributions to this repository must be signed as described on that page. Your signature certifies that you wrote the patch or have the right to pass it on as an open-source patch. For more detailed information, refer to [CONTRIBUTING.md](CONTRIBUTING.md).

# License
`terraform-provider-vra7` is available under the [MIT license](LICENSE).


CHANGELOG.md
## 0.2.0 (Unreleased)

FEATURES:

IMPROVEMENTS:

* Rename dirs/files according to the hashicorp provider's guidelines ([#145](https://github.com/vmware/terraform-provider-vra7/pull/145))

* Acceptance tests for vra7_deployment resource and fix for issue # 143 ([#144](https://github.com/vmware/terraform-provider-vra7/pull/144))


BUG FIXES:

* Acceptance tests for vra7_deployment resource and fix for issue # 143 ([#144](https://github.com/vmware/terraform-provider-vra7/pull/144))


## 0.1.0 (April 1, 2019)

FEATURES:

IMPROVEMENTS:

* Changes in the tf config file schema ([#135](https://github.com/vmware/terraform-provider-vra7/pull/135))

BUG FIXES:


## 0.0.2 (March 26, 2019)

FEATURES:

IMPROVEMENTS:

* Refactor code to split provider and SDK ([#119](https://github.com/vmware/terraform-provider-vra7/pull/119))

* Add more unit tests for the sdk and some refactoring ([#128](https://github.com/vmware/terraform-provider-vra7/pull/128))

BUG FIXES:

* Handle response pagination when fetching catalog item id by name ([#134](https://github.com/vmware/terraform-provider-vra7/pull/134))


## 0.0.1 (February 7, 2019)

FEATURES:

* Add requirement for go 1.11.4 or above ([#122](https://github.com/vmware/terraform-provider-vra7/issues/122))
* Convert from using dep to go modules ([#109](https://github.com/vmware/terraform-provider-vra7/issues/109))
* Adding businessgroup_name in the config file ([#94](https://github.com/vmware/terraform-provider-vra7/issues/94))
* Adding code to check if the component names in the terraform resource… ([#88](https://github.com/vmware/terraform-provider-vra7/issues/88))
* Get VM IP address ([#66](https://github.com/vmware/terraform-provider-vra7/issues/66))
* Update Deployment based on changes to configuration in Terraform file ([#27](https://github.com/vmware/terraform-provider-vra7/issues/27))
* resource_configuration key format verification check ([#36](https://github.com/vmware/terraform-provider-vra7/issues/36))
* Business Group Id resource field ([#28](https://github.com/vmware/terraform-provider-vra7/issues/28))
* Initial Pass at allowing 'description' and 'reasons' to be specified for a deployment ([#12](https://github.com/vmware/terraform-provider-vra7/issues/12))
* #7 Terraform "depends_on" does not wait - wait_timeout resource schema added. ([#10](https://github.com/vmware/terraform-provider-vra7/issues/10))
* Add insecure setting to allow connection with self-signed certs ([#3](https://github.com/vmware/terraform-provider-vra7/issues/3))

IMPROVEMENTS:

* Update README.md ([#114](https://github.com/vmware/terraform-provider-vra7/issues/114))
* Adding a logging framework for more detailed logging of vRA Terraform plugging. ([#85](https://github.com/vmware/terraform-provider-vra7/issues/85))
* Added debug messages to resource.go to help debug issues in the field. ([#80](https://github.com/vmware/terraform-provider-vra7/issues/80))
* Changes to variable and function names to better reflect vRA terminology ([#65](https://github.com/vmware/terraform-provider-vra7/issues/65))
* README.md changes ([#62](https://github.com/vmware/terraform-provider-vra7/issues/62))
* Unit testing - code coverage ([#48](https://github.com/vmware/terraform-provider-vra7/issues/48))
* Clean up the resource section of the README ([#32](https://github.com/vmware/terraform-provider-vra7/issues/32))
* Certificate signed by unknown authority README updates ([#16](https://github.com/vmware/terraform-provider-vra7/issues/16))
* Multi-machine blueprint terraform config example. ([#13](https://github.com/vmware/terraform-provider-vra7/issues/13))

BUG FIXES:

* Update go sum to fix the build failure ([#121](https://github.com/vmware/terraform-provider-vra7/issues/121))
* lease_days property name should be _leaseDays. ([#112](https://github.com/vmware/terraform-provider-vra7/issues/112))
* Have golint errors fail "make check" ([#108](https://github.com/vmware/terraform-provider-vra7/issues/108))
* Fix go lint errors/warnings ([#106](https://github.com/vmware/terraform-provider-vra7/issues/106))
* Cleanup travis tests ([#105](https://github.com/vmware/terraform-provider-vra7/issues/105))
* Fix terraform destroy. ([#103](https://github.com/vmware/terraform-provider-vra7/issues/103))
* Update issue templates ([#102](https://github.com/vmware/terraform-provider-vra7/issues/102))
* Correction in the schema ([#99](https://github.com/vmware/terraform-provider-vra7/issues/99))
* Fixing issues related to create, update and read ([#98](https://github.com/vmware/terraform-provider-vra7/issues/98))
* Fixing the config validation bug # 91 ([#92](https://github.com/vmware/terraform-provider-vra7/issues/92))
* Show request status on terraform update operation ([#90](https://github.com/vmware/terraform-provider-vra7/issues/90))
* Updating the request_status properly on time out. ([#86](https://github.com/vmware/terraform-provider-vra7/issues/86))
* merge crash fixes - minor change in add new value to machine config ([#64](https://github.com/vmware/terraform-provider-vra7/issues/64))
* Changes to the resource creation flow ([#55](https://github.com/vmware/terraform-provider-vra7/issues/55))
* Issue fix : Terraform destroy runs async (completes immediately) ([#56](https://github.com/vmware/terraform-provider-vra7/issues/56))
* Redo the error login in deleteResource to prevent panic ([#38](https://github.com/vmware/terraform-provider-vra7/issues/38))
* Add dynamic/deploy time properties appropriately from resource_configuration block ([#25](https://github.com/vmware/terraform-provider-vra7/issues/25))
* Use SplitN instead of Split to identify fields to replaces ([#29](https://github.com/vmware/terraform-provider-vra7/issues/29))
* Corrected minor typos in README.md ([#30](https://github.com/vmware/terraform-provider-vra7/issues/30))
* destroy resource outside terraform  error message fixes ([#22](https://github.com/vmware/terraform-provider-vra7/issues/22))


sdk/mock_data.go
package sdk

var (
	requestTemplateResponse = `{  
		"type":"com.vmware.vcac.catalog.domain.request.CatalogItemProvisioningRequest",
		"catalogItemId":"feaedf73-560c-4612-a573-41667e017691",
		"requestedFor":"fritz@coke.sqa-horizon.local",
		"businessGroupId":"b2470b94-cbca-43db-be37-803cca7b0f1a",
		"description":null,
		"reasons":null,
		"data":{  
		   "_leaseDays":null,
		   "_number_of_instances":1,
		   "machine2.vsphere":{  
			  "componentTypeId":"com.vmware.csp.component.cafe.composition",
			  "componentId":null,
			  "classId":"Blueprint.Component.Declaration",
			  "typeFilter":"Prativa_CentOs*machine2.vsphere",
			  "data":{  
				 "_cluster":1,
				 "_hasChildren":false,
				 "cpu":1,
				 "datacenter_location":null,
				 "description":null,
				 "disks":[  
					{  
					   "componentTypeId":"com.vmware.csp.iaas.blueprint.service",
					   "componentId":null,
					   "classId":"Infrastructure.Compute.Machine.MachineDisk",
					   "typeFilter":null,
					   "data":{  
						  "capacity":8,
						  "custom_properties":null,
						  "id":1537301083687,
						  "initial_location":"",
						  "is_clone":true,
						  "label":"Hard disk 1",
						  "storage_reservation_policy":"",
						  "userCreated":false,
						  "volumeId":0
					   }
					}
				 ],
				 "display_location":false,
				 "guest_customization_specification":null,
				 "location.loc":"",
				 "machine_prefix":null,
				 "max_network_adapters":-1,
				 "max_per_user":0,
				 "max_volumes":60,
				 "memory":1024,
				 "nics":null,
				 "os_arch":"x86_64",
				 "os_distribution":null,
				 "os_type":"Linux",
				 "os_version":null,
				 "ovfAuthNeeded":false,
				 "ovf_proxy_endpoint":null,
				 "ovf_url":null,
				 "ovf_url_pwd":null,
				 "ovf_url_username":null,
				 "ovf_use_proxy":false,
				 "property_groups":null,
				 "reservation_policy":null,
				 "security_groups":[  
	 
				 ],
				 "security_tags":[  
	 
				 ],
				 "snapshot_name":null,
				 "source_machine_external_snapshot":null,
				 "source_machine_vmsnapshot":null,
				 "storage":8
			  }
		   }
		}
	 }`

	requestTemplateErrorResponse = `{  
		"errors":[  
		   {  
			  "code":20116,
			  "source":null,
			  "message":"Unable to find the specified catalog item in the service catalog: feaedf73-560c-4612-a573-0041667e0176.",
			  "systemMessage":"Unable to find the specified catalog item in the service catalog: feaedf73-560c-4612-a573-0041667e0176.",
			  "moreInfoUrl":null
		   }
		]
	 }`

	catalogItemResp = `{  
		"catalogItem":{  
		   "callbacks":null,
		   "catalogItemTypeRef":{  
			  "id":"com.vmware.csp.component.cafe.composition.blueprint",
			  "label":"Composite Blueprint"
		   },
		   "dateCreated":"2015-12-22T03:16:19.289Z",
		   "description":"CentOS 6.3 IaaS Blueprint",
		   "forms":{  
			  "itemDetails":{  
				 "type":"external",
				 "formId":"composition.catalog.item.details"
			  },
			  "catalogRequestInfoHidden":true,
			  "requestFormScale":"BIG",
			  "requestSubmission":{  
				 "type":"extension",
				 "extensionId":"com.vmware.vcac.core.design.blueprints.requestForm",
				 "extensionPointId":null
			  },
			  "requestDetails":{  
				 "type":"extension",
				 "extensionId":"com.vmware.vcac.core.design.blueprints.requestDetailsForm",
				 "extensionPointId":null
			  },
			  "requestPreApproval":null,
			  "requestPostApproval":null
		   },
		   "iconId":"e5dd4fba-45ed-4943-b1fc-7f96239286be",
		   "id":"e5dd4fba-45ed-4943-b1fc-7f96239286be",
		   "isNoteworthy":false,
		   "lastUpdatedDate":"2017-01-06T05:12:56.690Z",
		   "name":"CentOS 6.3",
		   "organization":{  
			  "tenantRef":"vsphere.local",
			  "tenantLabel":"vsphere.local",
			  "subtenantRef":null,
			  "subtenantLabel":null
		   },
		   "outputResourceTypeRef":{  
			  "id":"composition.resource.type.deployment",
			  "label":"Deployment"
		   },
		   "providerBinding":{  
			  "bindingId":"vsphere.local!::!CentOS63",
			  "providerRef":{  
				 "id":"2fbaabc5-3a48-488a-9f2a-a42616345445",
				 "label":"Blueprint Service"
			  }
		   },
		   "serviceRef":{  
			  "id":"baad0ad2-8b96-4347-b188-f534dad53a0d",
			  "label":"Infrastructure"
		   },
		   "status":"PUBLISHED",
		   "statusName":"Published",
		   "quota":0,
		   "version":4,
		   "requestable":true
		},
		"entitledOrganizations":[  
		   {  
			  "tenantRef":"vsphere.local",
			  "tenantLabel":"vsphere.local",
			  "subtenantRef":"53619006-56bb-4788-9723-9eab79752cc1",
			  "subtenantLabel":"Content"
		   }
		]
	 }`

	entitledCatalogItemViewsResponse = `{  
		"links":[  
	 
		],
		"content":[  
		   {  
			  "@type":"ConsumerEntitledCatalogItemView",
			  "entitledOrganizations":[  
				 {  
					"tenantRef":"qe",
					"tenantLabel":"QETenant",
					"subtenantRef":"b2470b94-cbca-43db-be37-803cca7b0f1a",
					"subtenantLabel":"Development"
				 }
			  ],
			  "catalogItemId":"2e13fd45-a85e-4985-b89e-4ebb19ab272c",
			  "name":"NoSoftwareMachine",
			  "description":null,
			  "isNoteworthy":false,
			  "dateCreated":"2018-08-31T19:30:51.036Z",
			  "lastUpdatedDate":"2018-09-18T19:37:50.903Z",
			  "links":[  
				 {  
					"@type":"link",
					"rel":"GET: Request Template",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/2e13fd45-a85e-4985-b89e-4ebb19ab272c/requests/template"
				 },
				 {  
					"@type":"link",
					"rel":"POST: Submit Request",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/2e13fd45-a85e-4985-b89e-4ebb19ab272c/requests"
				 }
			  ],
			  "iconId":"composition.blueprint.png",
			  "catalogItemTypeRef":{  
				 "id":"com.vmware.csp.component.cafe.composition.blueprint",
				 "label":"Composite Blueprint"
			  },
			  "serviceRef":{  
				 "id":"d33afbc2-954d-416d-8d4a-bc4ad1b66058",
				 "label":"test-service"
			  },
			  "outputResourceTypeRef":{  
				 "id":"composition.resource.type.deployment",
				 "label":"Deployment"
			  }
		   },
		   {  
			  "@type":"ConsumerEntitledCatalogItemView",
			  "entitledOrganizations":[  
				 {  
					"tenantRef":"qe",
					"tenantLabel":"QETenant",
					"subtenantRef":"b2470b94-cbca-43db-be37-803cca7b0f1a",
					"subtenantLabel":"Development"
				 }
			  ],
			  "catalogItemId":"b7472041-d49b-44f2-a9b6-8ea714f55c55",
			  "name":"Dukes Bank Application",
			  "description":"Three-tier Dukes Bank application on Linux nodes; Apache HTTP Server for load balancer, JBoss Server for application server, and MySQL Server for database server. This is the basic version of Dukes Bank on Linux nodes.",
			  "isNoteworthy":false,
			  "dateCreated":"2018-08-31T18:12:49.680Z",
			  "lastUpdatedDate":"2018-09-18T19:24:09.673Z",
			  "links":[  
				 {  
					"@type":"link",
					"rel":"GET: Request Template",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/b7472041-d49b-44f2-a9b6-8ea714f55c55/requests/template"
				 },
				 {  
					"@type":"link",
					"rel":"POST: Submit Request",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/b7472041-d49b-44f2-a9b6-8ea714f55c55/requests"
				 }
			  ],
			  "iconId":"composition.blueprint.png",
			  "catalogItemTypeRef":{  
				 "id":"com.vmware.csp.component.cafe.composition.blueprint",
				 "label":"Composite Blueprint"
			  },
			  "serviceRef":{  
				 "id":"d33afbc2-954d-416d-8d4a-bc4ad1b66058",
				 "label":"test-service"
			  },
			  "outputResourceTypeRef":{  
				 "id":"composition.resource.type.deployment",
				 "label":"Deployment"
			  }
		   },
		   {  
			  "@type":"ConsumerEntitledCatalogItemView",
			  "entitledOrganizations":[  
				 {  
					"tenantRef":"qe",
					"tenantLabel":"QETenant",
					"subtenantRef":"b2470b94-cbca-43db-be37-803cca7b0f1a",
					"subtenantLabel":"Development"
				 }
			  ],
			  "catalogItemId":"1eb8e1d4-152e-4a93-a3b6-265df6870555",
			  "name":"Azure Machine (2)",
			  "description":"Creates new Azure virtual machine",
			  "isNoteworthy":false,
			  "dateCreated":"2018-09-18T19:40:19.172Z",
			  "lastUpdatedDate":"2018-09-18T19:41:39.432Z",
			  "links":[  
				 {  
					"@type":"link",
					"rel":"GET: Request Template",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/1eb8e1d4-152e-4a93-a3b6-265df6870555/requests/template"
				 },
				 {  
					"@type":"link",
					"rel":"POST: Submit Request",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/1eb8e1d4-152e-4a93-a3b6-265df6870555/requests"
				 }
			  ],
			  "iconId":"_internal!::!dda3cc3f-f2da-45f5-bc64-a292c1ace3ba_icon",
			  "catalogItemTypeRef":{  
				 "id":"com.vmware.csp.core.designer.service.serviceblueprint",
				 "label":"XaaS Blueprint"
			  },
			  "serviceRef":{  
				 "id":"d33afbc2-954d-416d-8d4a-bc4ad1b66058",
				 "label":"test-service"
			  },
			  "outputResourceTypeRef":{  
				 "id":"_internal!::!dda3cc3f-f2da-45f5-bc64-a292c1ace3ba",
				 "label":"Azure Virtual Machine"
			  }
		   },
		   {  
			  "@type":"ConsumerEntitledCatalogItemView",
			  "entitledOrganizations":[  
				 {  
					"tenantRef":"qe",
					"tenantLabel":"QETenant",
					"subtenantRef":"b2470b94-cbca-43db-be37-803cca7b0f1a",
					"subtenantLabel":"Development"
				 }
			  ],
			  "catalogItemId":"feaedf73-560c-4612-a573-41667e017691",
			  "name":"CentOs",
			  "description":"",
			  "isNoteworthy":false,
			  "dateCreated":"2018-09-18T20:04:55.805Z",
			  "lastUpdatedDate":"2018-09-27T22:52:48.390Z",
			  "links":[  
				 {  
					"@type":"link",
					"rel":"GET: Request Template",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/feaedf73-560c-4612-a573-41667e017691/requests/template"
				 },
				 {  
					"@type":"link",
					"rel":"POST: Submit Request",
					"href":"https://cava-n-80-072.eng.vmware.com/catalog-service/api/consumer/entitledCatalogItems/feaedf73-560c-4612-a573-41667e017691/requests"
				 }
			  ],
			  "iconId":"composition.blueprint.png",
			  "catalogItemTypeRef":{  
				 "id":"com.vmware.csp.component.cafe.composition.blueprint",
				 "label":"Composite Blueprint"
			  },
			  "serviceRef":{  
				 "id":"d33afbc2-954d-416d-8d4a-bc4ad1b66058",
				 "label":"test-service"
			  },
			  "outputResourceTypeRef":{  
				 "id":"composition.resource.type.deployment",
				 "label":"Deployment"
			  }
		   }
		],
		"metadata":{  
		   "size":20,
		   "totalElements":4,
		   "totalPages":1,
		   "number":1,
		   "offset":0
		}
	 }`

	subTenantsResponse = `{  
		"links":[  
	 
		],
		"content":[  
		   {  
			  "@type":"Subtenant",
			  "id":"b2470b94-cbca-43db-be37-803cca7b0f1a",
			  "name":"Development",
			  "description":"created by demo content",
			  "subtenantRoles":null,
			  "tenant":"qe",
			  "extensionData":{  
				 "entries":[  
					{  
					   "key":"iaas-manager-emails",
					   "value":{  
						  "type":"string",
						  "value":"astoyanov@vcac.sqa-horizon.local"
					   }
					}
				 ]
			  }
		   },
		   {  
			  "@type":"Subtenant",
			  "id":"ff371ec6-d4d8-4dee-aa73-f09e1bb3a4fd",
			  "name":"Quality Engineering",
			  "description":"created by demo content",
			  "subtenantRoles":null,
			  "tenant":"qe",
			  "extensionData":{  
				 "entries":[  
					{  
					   "key":"iaas-manager-emails",
					   "value":{  
						  "type":"string",
						  "value":"astoyanov@vcac.sqa-horizon.local"
					   }
					}
				 ]
			  }
		   },
		   {  
			  "@type":"Subtenant",
			  "id":"e11a782e-cff3-4707-b1e1-88940507dab3",
			  "name":"Finance",
			  "description":"created by demo content",
			  "subtenantRoles":null,
			  "tenant":"qe",
			  "extensionData":{  
				 "entries":[  
					{  
					   "key":"iaas-manager-emails",
					   "value":{  
						  "type":"string",
						  "value":"gloria@coke.sqa-horizon.local"
					   }
					}
				 ]
			  }
		   },
		   {  
			  "@type":"Subtenant",
			  "id":"6273f684-210c-4839-85ec-573346f53799",
			  "name":"cloudclient-ITs-bg",
			  "description":"",
			  "subtenantRoles":null,
			  "tenant":"qe",
			  "extensionData":{  
				 "entries":[  
					{  
					   "key":"iaas-machine-prefix",
					   "value":{  
						  "type":"string",
						  "value":"b272b729-1583-4e6e-b299-15c293a6d0fa"
					   }
					},
					{  
					   "key":"iaas-ad-container",
					   "value":{  
						  "type":"string",
						  "value":""
					   }
					},
					{  
					   "key":"iaas-manager-emails",
					   "value":{  
						  "type":"string",
						  "value":"fritz@sqa-local.com"
					   }
					}
				 ]
			  }
		   }
		],
		"metadata":{  
		   "size":20,
		   "totalElements":4,
		   "totalPages":1,
		   "number":1,
		   "offset":0
		}
	 }`
)
